TaylorFramework/Modules/temper/Output/PMDCoordinator.swift

```diff
@@ -15,36 +15,41 @@ final class PMDCoordinator: WritingCoordinator {
         let xml = generateXML(violations)
         let xmlData = xml.xmlData(withOptions: Int(XMLNode.Options.nodePrettyPrint.rawValue))
         do {
-            try xmlData.write(to: URL(fileURLWithPath: path), options: NSData.WritingOptions.withoutOverwriting)
-        } catch {
-            print("Error while writing the XML object to file.")
+            try xmlData.write(to: URL(fileURLWithPath: path), options: .atomic)
+        } catch let error {
+            let message = "Error while writing XML object at path: " + path + "\n" +
+                          "Reason: " + error.localizedDescription
+            Printer(verbosityLevel: .error).printError(message)
         }
     }
     
-    fileprivate func generateXML(_ violations: [Violation]) -> XMLDocument {
+    private func generateXML(_ violations: [Violation]) -> XMLDocument {
         let xml = XMLDocument(rootElement: XMLElement(name: "pmd"))
         xml.version = "1.0"
         xml.characterEncoding = "UTF-8"
-        addElementsToXML(xml, fromViolations: violations)
+        addElements(to: xml, from: violations)
         return xml
     }
     
-    fileprivate func addElementsToXML(_ xml: XMLDocument, fromViolations violations: [Violation]) {
-        for filePath in filePathsFromViolations(violations) {
-            let fileElement = XMLElement(name: "file")
-            let attributeNode = XMLNode.attribute(withName: "name", stringValue: filePath) as? XMLNode
-            guard let attribute = attributeNode else {
-                return
-            }
-            fileElement.addAttribute(attribute)
-            let _ = violations.filter({ $0.path == filePath }).map({ fileElement.addChild($0.toXMLElement()) })
-            if let root = xml.rootElement() {
-                root.addChild(fileElement)
-            }
+    private func addElements(to xml: XMLDocument, from violations: [Violation]) {
+        for path in filePaths(from: violations) {
+            guard let fileElement = xmlElement(withFilePath: path) else { continue }
+            violations.filter { $0.path == path }
```

Comment:
It seems like we do a lot of iteration here. A big number of violations can lead to a computation time growth.
One way we can reduce the numbers of loops is by iterating through violations (just once), group them into a dictionary that uses paths as keys:
```swift
private typealias ViolationsMap = [Path: [Violation]]

final class PMDCoordinator: WritingCoordinator {
    
    func writeViolations(_ violations: [Violation], atPath path: String) {
        FileManager().removeFileAtPath(path)
        let map = mapViolations(violations)
        let xml = generateXML(from: map)
        let xmlData = xml.xmlData(withOptions: Int(XMLNode.Options.nodePrettyPrint.rawValue))
        do {
            try xmlData.write(to: URL(fileURLWithPath: path), options: .atomic)
        } catch let error {
            let message = "Error while writing XML object at path: " + path + "\n" +
                "Reason: " + error.localizedDescription
            Printer(verbosityLevel: .error).printError(message)
        }
    }
    
    private func mapViolations(_ violations: [Violation]) -> ViolationsMap {
        return violations.reduce([Path : [Violation]]()) { result, violation in
            let path = violation.path
            
            // Add violation to the group with the same path
            var violations = result[path] ?? []
            violations.append(violation)
            
            // Update the result
            var nextResult = result
            nextResult[path] = violations
            
            return result
        }
    }
    
    private func generateXML(from violationsMap: ViolationsMap) -> XMLDocument {
        let xml = XMLDocument(rootElement: XMLElement(name: "pmd"))
        xml.version = "1.0"
        xml.characterEncoding = "UTF-8"
        
        let fileElements = violationsMap.map(generateFileElement)
        xml.rootElement()?.addChildren(fileElements)
        
        return xml
    }
    
    private func generateFileElement(path: Path, violations: [Violation]) -> XMLElement {
        let element = XMLElement(name: "file")
        let attribute = XMLNode(kind: .attribute)
        attribute.setValue(path, forKey: "name")
        element.addAttribute(attribute)
        
        violations.forEach { element.addChild($0.toXMLElement()) }
        
        return element
    }
}

extension XMLElement {
    func addChildren(_ children: [XMLElement]) {
        children.forEach { child in
            addChild(child)
        }
    }
}
```
What do you think?